### ISSUE TITLE
PnL reports "incomplete cost basis for spend" on trades based on profit currency

### ISSUE THREAD

**Issue by idelgado** [rotki/rotki#3947]
## Problem Definition

The PnL report throws the following message when computing cost basis related to 5 USD->ADA buys and 1 ADA->SOL sell. Investigating further it looks like the issue may be related to rates derived from the price oracle rather than being derived from the trades themselves.

> Not enough documented acquisitions found for ADA(Cardano) before 12/10/2021 08:23:56. Only found acquisitions for 37004.53663936611576641883254 ADA(Cardano) and miss 729.49792163388423358116746 ADA(Cardano).Let rotki know how you acquired it via a ledger action

> Incomplete cost basis information for spend. Used: 2677.374566366115766418832538 / 2951.286057 acquired in coinbase at 15/05/2021 11:59:18|4123.161526 / 4123.161526 acquired in coinbase at 19/05/2021 11:13:25|4125.287831 / 4125.287831 acquired in coinbase at 20/05/2021 07:22:16|26078.712716 / 26078.712716 acquired in coinbase at 20/05/2021 07:32:58

| type | location | paid_asset                | paid_in_asset                 | taxable_amount | received_asset | received_in_asset             | net_profit_or_loss      | time                | is_virtual | paid_in_USD                   | taxable_received_in_USD       | taxable_bought_cost_in_USD    | cost_basis                                                                                                                                                                                                                                                                                                                                                        | total_bought_cost_in_USD      | total_received_in_USD         | link                                 | notes | 
|------|----------|---------------------------|-------------------------------|----------------|----------------|-------------------------------|-------------------------|---------------------|------------|-------------------------------|-------------------------------|-------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|-------------------------------|-------------------------------|--------------------------------------|-------| 
| buy  | coinbase | USD(United States Dollar) | 1050.540000000000000000000000 | 0              | ADA(Cardano)   | 455.586431                    | 0                       | 15/05/2021 11:56:48 | FALSE      | 1066.190000000000000000000000 | 0                             | 0                             |                                                                                                                                                                                                                                                                                                                                                                   | 0                             | 0                             | aecb3909-42d1-53b8-98fd-9e6b5ba9db26 |       | 
| buy  | coinbase | USD(United States Dollar) | 6897.230000000000000000000000 | 0              | ADA(Cardano)   | 2951.286057                   | 0                       | 15/05/2021 11:59:18 | FALSE      | 7000.000000000000000000000000 | 0                             | 0                             |                                                                                                                                                                                                                                                                                                                                                                   | 0                             | 0                             | 20cb0018-08fd-5525-8497-79abe0d5da98 |       | 
| buy  | coinbase | USD(United States Dollar) | 6897.230000000000000000000000 | 0              | ADA(Cardano)   | 4123.161526                   | 0                       | 19/05/2021 11:13:25 | FALSE      | 7000.000000000000000000000000 | 0                             | 0                             |                                                                                                                                                                                                                                                                                                                                                                   | 0                             | 0                             | 46503731-b033-517c-b9db-36e7cb6fc7ed |       | 
| buy  | coinbase | USD(United States Dollar) | 7640.920000000000000000000000 | 0              | ADA(Cardano)   | 4125.287831                   | 0                       | 20/05/2021 07:22:16 | FALSE      | 7754.770000000000000000000000 | 0                             | 0                             |                                                                                                                                                                                                                                                                                                                                                                   | 0                             | 0                             | 559d2b26-e4dc-5ac6-a3bc-e09777653e48 |       | 
| buy  | coinbase | USD(United States Dollar) | 49265.94000000000000000000000 | 0              | ADA(Cardano)   | 26078.712716                  | 0                       | 20/05/2021 07:32:58 | FALSE      | 50000.00000000000000000000000 | 0                             | 0                             |                                                                                                                                                                                                                                                                                                                                                                   | 0                             | 0                             | 186ae115-5aa7-5b50-89a5-2af158171d00 |       | 
| sell | coinbase | ADA(Cardano)              | 37734.034561                  | 37734.034561   | SOL(Solana)    | 552.3158858360000000000000000 | =IF(E218=0;0;L218-M218) | 12/10/2021 08:23:56 | FALSE      | 0                             | 76645.34552515690000000000000 | 71105.09375804800000000000000 | Incomplete cost basis information for spend. Used: 2677.374566366115766418832538 / 2951.286057 acquired in coinbase at 15/05/2021 11:59:18\|4123.161526 / 4123.161526 acquired in coinbase at 19/05/2021 11:13:25\|4125.287831 / 4125.287831 acquired in coinbase at 20/05/2021 07:22:16\|26078.712716 / 26078.712716 acquired in coinbase at 20/05/2021 07:32:58 | 72648.34661126450000000000000 | 78188.59837837330000000000000 | d97ccd41-13ab-51b7-83d1-2ef0db3e7a03 |       | 
| buy  | coinbase | ADA(Cardano)              | 37734.03456100000000000000000 | 0              | SOL(Solana)    | 552.3158858360000000000000000 | 0                       | 12/10/2021 08:23:56 | TRUE       | 79826.35011379550000000000000 | 0                             | 0                             |                                                                                                                                                                                                                                                                                                                                                                   | 0                             | 0                             | d97ccd41-13ab-51b7-83d1-2ef0db3e7a03 |       | 

## Logs

> [15/01/2022 19:34:15 CST] DEBUG rotkehlchen.history.price Greenlet-6: Querying historical price from_asset=SOL(Solana), to_asset=USD(United States Dollar), timestamp=1634045036
[15/01/2022 19:34:15 CST] DEBUG rotkehlchen.externalapis.cryptocompare Greenlet-6: Will query Cryptocompare history for SOL-2 -> USD @ 1634045036. Cached data: True rate_limited in last 60 seconds: False
[15/01/2022 19:34:15 CST] DEBUG rotkehlchen.externalapis.cryptocompare Greenlet-6: Got historical price from cryptocompare from_asset=SOL(Solana), to_asset=USD(United States Dollar), timestamp=1634045036, price=0
[15/01/2022 19:34:15 CST] DEBUG rotkehlchen.history.price Greenlet-6: Historical price oracle cryptocompare got price price=141.565, from_asset=SOL(Solana), to_asset=USD(United States Dollar), timestamp=1634045036
[15/01/2022 19:34:15 CST] DEBUG rotkehlchen.history.price Greenlet-6: Querying historical price from_asset=ADA(Cardano), to_asset=USD(United States Dollar), timestamp=1634045036
[15/01/2022 19:34:15 CST] DEBUG rotkehlchen.externalapis.cryptocompare Greenlet-6: Will query Cryptocompare history for ADA -> USD @ 1634045036. Cached data: True rate_limited in last 60 seconds: False
[15/01/2022 19:34:15 CST] DEBUG rotkehlchen.externalapis.cryptocompare Greenlet-6: Got historical price from cryptocompare from_asset=ADA(Cardano), to_asset=USD(United States Dollar), timestamp=1634045036, price=0
[15/01/2022 19:34:15 CST] DEBUG rotkehlchen.history.price Greenlet-6: Historical price oracle cryptocompare got price price=2.1155, from_asset=ADA(Cardano), to_asset=USD(United States Dollar), timestamp=1634045036
[15/01/2022 19:34:15 CST] DEBUG rotkehlchen.accounting.events Greenlet-6: Processing sell trade of ADA for SOL-2 at 1634045036
[15/01/2022 19:34:15 CST] DEBUG rotkehlchen.accounting.cost_basis Greenlet-6: AssetSpendEvent in coinbase @ 1634045036.amount: 37734.034561 rate: 2.072097492039325744126330557 fee_rate: 0.04089816716316655456462785440 gain_in_profit_currency: 78188.59837837333999999999999
[15/01/2022 19:34:15 CST] DEBUG rotkehlchen.accounting.cost_basis Greenlet-6: Spend uses up entire historical acquisition tax_status=TAXABLE, bought_amount=455.586431, asset=ADA(Cardano), acquisition_rate=2.305907130934722680535672056, profit_currency=USD(United States Dollar), time=15/05/2021 11:56:48
[15/01/2022 19:34:15 CST] DEBUG rotkehlchen.accounting.cost_basis Greenlet-6: Spend uses up part of historical acquisition tax_status=TAXABLE, used_amount=273.9114906338842335811674623, from_amount=2951.286057, asset=ADA(Cardano), acquisition_rate=2.337025238079115825945163539, profit_currency=USD(United States Dollar), time=15/05/2021 11:59:18
[15/01/2022 19:34:15 CST] DEBUG rotkehlchen.accounting.cost_basis Greenlet-6: Spend uses up entire historical acquisition tax_status=TAXABLE, bought_amount=2677.374566366115766418832538, asset=ADA(Cardano), acquisition_rate=2.337025238079115825945163539, profit_currency=USD(United States Dollar), time=15/05/2021 11:59:18
[15/01/2022 19:34:15 CST] DEBUG rotkehlchen.accounting.cost_basis Greenlet-6: Spend uses up entire historical acquisition tax_status=TAXABLE, bought_amount=4123.161526, asset=ADA(Cardano), acquisition_rate=1.672801309506592441947422266, profit_currency=USD(United States Dollar), time=19/05/2021 11:13:25
[15/01/2022 19:34:15 CST] DEBUG rotkehlchen.accounting.cost_basis Greenlet-6: Spend uses up entire historical acquisition tax_status=TAXABLE, bought_amount=4125.287831, asset=ADA(Cardano), acquisition_rate=1.852215000025291568558189170, profit_currency=USD(United States Dollar), time=20/05/2021 07:22:16
[15/01/2022 19:34:15 CST] DEBUG rotkehlchen.accounting.cost_basis Greenlet-6: Spend uses up entire historical acquisition tax_status=TAXABLE, bought_amount=26078.712716, asset=ADA(Cardano), acquisition_rate=1.889124687115940542806967282, profit_currency=USD(United States Dollar), time=20/05/2021 07:32:58
[15/01/2022 19:34:15 CST] ERROR rotkehlchen.user_messages Greenlet-6: Not enough documented acquisitions found for ADA(Cardano) before 12/10/2021 08:23:56. Only found acquisitions for 37004.53663936611576641883254 ADA(Cardano) and miss 729.49792163388423358116746 ADA(Cardano).Let rotki know how you acquired it via a ledger action
[15/01/2022 19:34:15 CST] DEBUG rotkehlchen.accounting.events Greenlet-6: After Sell Profit/Loss taxable_profit_loss=5540.25176710888960929265968, general_profit_loss=5540.25176710888960929265968, profit_currency=USD(United States Dollar)

<img width="742" alt="Rotki Rates" src="https://user-images.githubusercontent.com/882552/149674632-9eb22ebf-ddd2-4b63-99e3-1b2ba8835dba.png">

<img width="466" alt="cointracker.io rates" src="https://user-images.githubusercontent.com/882552/149674631-a624cbb8-32b0-492d-a794-becbb5b0676b.png">

[related_events.csv](https://github.com/rotki/rotki/files/7877736/related_events.csv)

### System Description


Operating system: macOS Big Sur 11.6 
Rotki version: 1.23.1

**Comment by idelgado** [comment]
I think I made sense of the error notification reported in the app:
> Not enough documented acquisitions found for ADA(Cardano) before 12/10/2021 08:23:56. Only found acquisitions for 37004.53663936611576641883254 ADA(Cardano) and miss 729.49792163388423358116746 ADA(Cardano).Let rotki know how you acquired it via a ledger action

The 729.49792163388423358116746 ADA(Cardano) corresponds to the fee associated with the trade from ADA to SOL. This is displayed in the dropdown of the trade in the app as a fee.

I'm still trying to make sense of the other message related to cost basis.

**Comment by idelgado** [comment]
I determined that the cost basis error stems from the fact that acquisitions are deducted for the fee_amount. However, when computing the remaining acquisitions this fee_amount is not deducted from the selling_amount. This causes the acquisition to be short by the fee_amount as described in the error message.

I tried the following and for this case it seemed to have resolved the issue. I'm still validating this atm but thought i'd share what i've found so far.

```diff
diff --git a/rotkehlchen/accounting/events.py b/rotkehlchen/accounting/events.py
index 1892df4fe..28fddfc01 100644
--- a/rotkehlchen/accounting/events.py
+++ b/rotkehlchen/accounting/events.py
@@ -583,6 +583,9 @@ class TaxableEvents():
                 spending_asset=fee_currency,
                 timestamp=timestamp,
             )
+            if fee_currency == selling_asset:
+                # deduct the fee_amount
+                selling_amount = selling_amount - fee_amount
         self.handle_prefork_asset_sells(selling_asset, selling_amount, timestamp)
 
         # now search the acquisitions for `paid_with_asset` and calculate profit/loss
```

**Comment by yabirgb** [comment]
> I determined that the cost basis error stems from the fact that acquisitions are deducted for the fee_amount. However, when computing the remaining acquisitions this fee_amount is not deducted from the selling_amount. This causes the acquisition to be short by the fee_amount as described in the error message:
> 
> > Not enough documented acquisitions found for ADA(Cardano) before 12/10/2021 08:23:56. Only found acquisitions for 37004.53663936611576641883254 ADA(Cardano) and miss 729.49792163388423358116746 ADA(Cardano).Let rotki know how you acquired it via a ledger action
> 
> I tried the following and for this case it seemed to have resolved the issue. I'm still validating this atm but thought i'd share what i've found so far.
> 
> ```diff
> diff --git a/rotkehlchen/accounting/events.py b/rotkehlchen/accounting/events.py
> index 1892df4fe..28fddfc01 100644
> --- a/rotkehlchen/accounting/events.py
> +++ b/rotkehlchen/accounting/events.py
> @@ -583,6 +583,9 @@ class TaxableEvents():
>                  spending_asset=fee_currency,
>                  timestamp=timestamp,
>              )
> +            if fee_currency == selling_asset:
> +                # deduct the fee_amount
> +                selling_amount = selling_amount - fee_amount
>          self.handle_prefork_asset_sells(selling_asset, selling_amount, timestamp)
>  
>          # now search the acquisitions for `paid_with_asset` and calculate profit/loss
> ```

Hola @idelgado ! The situation you want to fix in this comment was recently handled by https://github.com/rotki/rotki/pull/3933/ . 
As you said before the error might be related to the price we choose at the moment of calculating the trades in the native currency since we query oracles instead of checking the trade. I'm preparing a fix that I need to test a bit and we will hang you if possible a binary for testing is that is fine for you. Working on this issue now

**Comment by yabirgb** [comment]
Wrote you a DM on discord @idelgado !

**Comment by LefterisJP** [comment]
Should be fixed by #4029
